tests/Struct/FindologicProductTest.php

```diff
@@ -366,6 +372,375 @@ public function testProduct(): void
+        );
+
+        $findologicImages = $findologicProduct->getImages();
+        $this->assertEquals($images, $findologicImages);
+    }
+
+    /**
+     * @return Image[]
+     */
+    private function getImages(ProductEntity $productEntity): array
+    {
+        $images = [];
+        $media = $productEntity->getMedia();
+        if (!$media || !$media->count()) {
+            $requestContext = $this->router->getContext();
+            $schemaAuthority = $requestContext->getScheme() . '://' . $requestContext->getHost();
+            if ($requestContext->getHttpPort() !== 80) {
+                $schemaAuthority .= ':' . $requestContext->getHttpPort();
+            } elseif ($requestContext->getHttpsPort() !== 443) {
+                $schemaAuthority .= ':' . $requestContext->getHttpsPort();
+            }
+
+            $fallbackImage = sprintf(
+                '%s/%s',
+                $schemaAuthority,
+                'bundles/storefront/assets/icon/default/placeholder.svg'
+            );
+
+            $images[] = new Image($fallbackImage);
+            $images[] = new Image($fallbackImage, Image::TYPE_THUMBNAIL);
+
+            return $images;
+        }
+
+        foreach ($media as $mediaEntity) {
+            $media = $mediaEntity->getMedia();
+            if (!$media || !$media->getUrl()) {
+                continue;
+            }
+
+            $thumbnails = $media->getThumbnails();
+            if (!$thumbnails) {
+                $encodedUrl = $this->getEncodedUrl($media->getUrl());
+                if (!Utils::isEmpty($encodedUrl)) {
+                    $images[] = new Image($encodedUrl);
+                }
+                continue;
+            }
+
+            $filteredThumbnails = $this->filterThumbnails($thumbnails);
+            $image = $filteredThumbnails->first() ?? $media;
+            if ($image) {
+                $encodedThumbnailUrl = $this->getEncodedUrl($image->getUrl());
+                if (!Utils::isEmpty($encodedThumbnailUrl)) {
+                    $images[] = new Image($encodedThumbnailUrl);
+                }
+            }
+
+            foreach ($thumbnails as $thumbnailEntity) {
+                $encodedThumbnailUrl = $this->getEncodedUrl($thumbnailEntity->getUrl());
+                if (!Utils::isEmpty($encodedThumbnailUrl)) {
+                    $images[] = new Image($encodedThumbnailUrl, Image::TYPE_THUMBNAIL);
+                }
+            }
+        }
+
+        return $images;
+    }
+
+    protected function getEncodedUrl(string $url): string
+    {
+        $parsedUrl = parse_url($url);
+        $urlPath = explode('/', $parsedUrl['path']);
+        $encodedPath = array_map('\FINDOLOGIC\FinSearch\Utils\Utils::multiByteRawUrlEncode', $urlPath);
+        $parsedUrl['path'] = implode('/', $encodedPath);
+
+        return Utils::buildUrl($parsedUrl);
+    }
+
+    /**
+     * @return Attribute[]
+     */
+    private function getAttributes(ProductEntity $productEntity): array
```

Comment:
Why are methods moved everywhere? Please move them back to where they were in the first place. It is not just hard to review but also not really beneficial to anyone.

Reply:
Actually the IDE does that to sort related methods together. I have disabled that setting. Won't happen anymore.

Reply:
Yes please. Reordering methods just makes it really hard to review.